### ISSUE TITLE
Saved microphone permissions seemingly not picked up on Firefox

### ISSUE THREAD

**Issue by msub2** [speechly/speech-recognition-polyfill#7]
### Describe the bug
The polyfill doesn't appear to transmit audio data in Firefox if microphone permissions have been set to 'Allow' and remembered.

### To Reproduce
1. Create a basic NPM project with the speech recognition polyfill integrated. In my case I made a minimal HTML page that calls `start()` when you press space and `stop()` when you release it. The transcript from the result is then printed by the `onresult` handler.
2. Call `start()` and speak a test utterance.
3. Call `stop()` and wait for the transcript to be printed.

### Expected behaviour
When I call stop (release the spacebar) I will receive a transcript of what I said.

### Actual behaviour
Nothing is printed, even though the network tab in the developer tools shows that messages are being sent.

### Environment

- Platform: Desktop
- OS: Windows 10
- Browser: Firefox
- Version: 92.0
- Package version: 1.0.0

### Additional context
The only way I've gotten it to function correctly is by explicitly granting it permissions every time when it asks, which appears to be when `start()` is called for the first time. Here is the index.js file from my test project:
```js
import { createSpeechlySpeechRecognition } from '@speechly/speech-recognition-polyfill';

const appId = '<my-app-id>';
const SpeechlySpeechRecognition = createSpeechlySpeechRecognition(appId);
const speechRecognition = new SpeechlySpeechRecognition();

let speaking = false;

document.addEventListener('keydown', e => {
  if (e.code === 'Space') {
    if (!speaking) {
      console.log('started speaking');
      speaking = true;
      speechRecognition.start();
    }
  }
});
document.addEventListener('keyup', e => {
  if (e.code === 'Space') {
    if (speaking) {
      console.log('finished speaking');
      speaking = false;
      speechRecognition.stop();
    }
  }
});
speechRecognition.onresult = e => {
  console.log(e.results[0][0].transcript);
};

```

I get the feeling this might even be an upstream bug with the browser client, but I haven't dug into it yet so I'm not totally sure yet. I have not observed this behavior in any Chromium-based browsers (tested in Chrome and Oculus Browser).

**Comment by JamesBrill** [comment]
Hi @msub2 Thank you for raising this issue and filling the template in such detail.

Unfortunately, this has worked as expected for me so far. I copied your script exactly and tested its behaviour in Firefox 92 (on OS X as I don't have access to Windows at the moment). I've captured the behaviour I observed in the GIF below:
* Press spacebar
* Give microphone permission and ask Firefox to remember the preference
* Speak and observe the transcript being printed
* Refresh the page
* Speak again and observe another transcript

![Sep-21-2021 21-40-01](https://user-images.githubusercontent.com/2140027/134244441-869c3ff1-4dd9-4ec2-9f43-9c044bdbc23b.gif)

I was also able to get the desired behaviour when opening a second tab and continuing there.

Note that the emission of the transcript will not be tied to the release of the spacebar. The transcript will be emitted when the speech recognition engine thinks you've finished speaking, which may occur before or after the spacebar is released.

Also note that, by default, the speech recognition engine will only transcribe one utterance for each call to `start()`. So if you get a transcript while holding the spacebar down, you won't get another one until you've released it and pressed it again. If you want multiple utterances to be transcribed for one key press, then you can set the `continuous` property to `true` on the recognition object.

I don't have any ideas to unblock you yet, though these questions may help me understand the behaviour you're seeing better:
* If you select Allow _without_ remembering the choice, does the transcription work as expected? Does it continue to work when refreshing the same tab?
* When you select Allow and remember the choice, does it stop working completely, even when opening other tabs or refreshing? Or does it work for the first tab and fail for the others?
* While the transcription is not working, is there any indicator that the microphone is turned on? i.e. is it the microphone that's the issue rather than the transcription?
* Is it possible that you have an extension or non-standard configuration in Firefox that is affecting the microphone or network behaviour?
* Are you running on `localhost`?

**Comment by msub2** [comment]
Thanks for taking a look James. To answer your questions:

- Yes, I am seeing similar behavior on that Speechly demo.
- Choosing allow *without* remembering the choice will consistently work.
- Allowing and remembering fails on the first tab, even after refreshing.
- I am confident that the microphone is working correctly.
- I suppose this isn't impossible, but I don't know of any of my extensions that would be affecting it like this.
- This behavior is failing on localhost.

I've also recorded a short 2 minute video showing the behavior from my side. https://streamable.com/d8cwpc

**Comment by JamesBrill** [comment]
Thanks for answering the questions - that video is especially useful. It's getting late here in the UK, so I'll have to return to this later in the week, perhaps when I have access to my Windows machine.

I do have one more idea though. I notice you have multiple audio devices. From my own experience, I know that Windows can sometimes be inconsistent in how it chooses audio devices for each application, even when you've defined your preferred defaults. I wonder if you're asking Firefox to remember Device A, but either Windows or Firefox then tries to use Device B when you refresh the page, perhaps because Device B is a default device. I'm not sure why that would break transcription, but perhaps there's some kind of audio encoding mismatch in that case - Firefox tells the Speechly browser client via the MediaDevices API to expect audio from Device A, but it gets audio from Device B instead. Or maybe Firefox's implementation of the MediaDevices API has a bug that ignores your chosen audio device and just uses a default one every time.

It's a bit of wild speculation, but it's the only other noticeable difference between your setup and mine - I only have one input device, so there's no possibility of a mismatch.

So there's a couple of tests you could do:
* Perform the same process from your video on another website that requires microphone permission (e.g. [Bing search](https://www.bing.com/)), one unrelated to Speechly. If you experience it again there, this could be a bug in Firefox (or its implementation of the MediaDevices API) and/or Windows
* Remove all your audio devices except for one so that Windows only ever has one to choose from. Then repeat what you did in the video

**Comment by msub2** [comment]
Your second suggestion seems to have done it! It looks like Firefox was indeed just grabbing the first microphone from that list, which was the virtual mic input used by Oculus for Oculus Link, hence no sound. After disabling all but my actual microphone input things seem to be working correctly now when I choose to remember permissions.

**Comment by JamesBrill** [comment]
Great to hear! Although this seems like a Firefox bug, the team at Speechly are looking into whether the browser client can be modified to detect this edge case.

Given that's in a different repo and there's no action that can be taken within this polyfill repo, I'm closing this issue. If you want to keep tabs on this, you're welcome to raise a new issue in the [browser client](https://github.com/speechly/browser-client) repo, referencing this one if needed.
